packages/connection-form/src/components/advanced-options-tabs/csfle-tab/encrypted-field-config-input.tsx

```diff
@@ -0,0 +1,98 @@
+import React from 'react';
+import type { Document } from 'mongodb';
+import {
+  encryptedFieldConfigToText,
+  textToEncryptedFieldConfig,
+} from '../../../utils/csfle-handler';
+import {
+  Label,
+  Banner,
+  Description,
+  css,
+  spacing,
+} from '@mongodb-js/compass-components';
+
+import 'ace-builds';
+import AceEditor from 'react-ace';
+import 'ace-builds/src-noconflict/ext-language_tools';
+import 'mongodb-ace-mode';
+import 'mongodb-ace-theme';
+
+const errorContainerStyles = css({
+  padding: spacing[3],
+  width: '100%',
+});
+
+const withMarginStyles = css({
+  marginTop: spacing[1],
+});
+
+/**
+ * Options for the ACE editor.
+ */
+const OPTIONS = {
+  enableLiveAutocompletion: false,
+  tabSize: 2,
+  fontSize: 11,
+  minLines: 10,
+  maxLines: Infinity,
+  showGutter: true,
+  useWorker: false,
+  mode: 'ace/mode/mongodb',
+};
+
+function EncryptedFieldConfigInput({
+  encryptedFieldConfig,
+  errorMessage,
+  onChange,
+}: {
+  encryptedFieldConfig: Document | undefined;
+  errorMessage: string | undefined;
+  onChange: (value: Document | undefined) => void;
+}): React.ReactElement {
+  function changeCurrentText(newText: string) {
+    if (!newText.trim()) {
+      onChange(undefined);
+      return;
+    }
+
+    onChange(textToEncryptedFieldConfig(newText));
+  }
+
+  const currentText = encryptedFieldConfig
+    ? encryptedFieldConfigToText(encryptedFieldConfig)
+    : '';
+
+  return (
+    <div>
+      <div className={withMarginStyles}>
+        <Label htmlFor="TODO(COMPASS-5653)">EncryptedFieldConfigMap</Label>
+      </div>
+      <div className={withMarginStyles}>
+        <Description className={withMarginStyles}>
+          Add an optional client-side EncryptedFieldConfigMap for enhanced
+          security.
+        </Description>
+      </div>
+      <div className={withMarginStyles}>
+        <AceEditor
+          mode="javascript" // will be set to mongodb as part of OPTIONS
+          theme="mongodb"
+          width="100%"
+          value={currentText}
+          onChange={changeCurrentText}
```

Comment:
I'm receiving an error when typing in this field:
![Screen Shot 2022-04-01 at 12 44 31 PM](https://user-images.githubusercontent.com/1791149/161307107-94497aed-efeb-414a-a723-1504c9d4db85.png)

Is that expected for now?

Reply:
Since this action does some more logic with the `textToEncryptedFieldConfig` function do we want to add a test for this component that types into this editor and spies on the response to the `onChange` prop we pass to the component?
An example of a test like this: https://github.com/mongodb-js/compass/blob/main/packages/compass-find-in-page/src/components/find-in-page-input.spec.tsx#L121 - there are also probably a few in this package to reference as well.

Reply:
Ugh, thanks for noticing this error. No, it’s not expected, it’s a weird quirk of our `ejson-shell-parser` package that I cannot understand at all, where invalid input is mapped to an empty string as a result, rather than an exception.

(This is worrying me a little bit, because it means that we might be reading an empty string where we expect something else in other places as well. Here this would have been particularly bad, since accidentally omitting a clien-provided `EncryptedFieldConfig` can be dangerous from a security perspective.)

> Since this action does some more logic with the `textToEncryptedFieldConfig` function do we want to add a test for this component that types into this editor and spies on the response to the `onChange` prop we pass to the component?

I’ve tinkered around a bit with writing a test like this, but in the end I’m not sure how to make this work with the ace editor rather than a text input.

Fwiw, the bulk of the work here is done by `textToEncryptedFieldConfig()` and `encryptedFieldConfigToText()`, and those are tested quite well, so I’ve moved the remainder of the text conversions there as well.

---

Somewhat relatedly, I noticed that we also need to use the logic here for reviving EncryptedFieldConfig when double-clicking through the sidebar, where otherwise the stored data in the model would have led to errors. I’ve exposed a wrapper around `adjustCSFLEParams()` in the connection-form exports (see 5f8f8197a99e8def5770fdfb76ec3a181871b9ac), it doesn’t feel like an ideal solution here though and I’m happy to take suggestions on where a better place for this would be.

Reply:
Gotcha - nice find. 

For the location of `adjustCSFLEParams` - would it make sense to be in the data-service instead?
Something on load/save? I'm thinking then the connect layer would only interact with the expected connectionOptions.
https://github.com/mongodb-js/compass/blob/main/packages/data-service/src/legacy/legacy-connection-model.ts#L206
Used here: https://github.com/mongodb-js/compass/blob/main/packages/data-service/src/connection-storage.ts#L46
Not sure if that's a better spot, just a thought. I think ideally we want this connection form to be usable in places like VSCode or other projects down the line - it would make sense to keep the connection form logic/usage to be minimal where we can then.

Reply:
Mhh … I thought about that, and actually wrote a version that placed it in data-service at some point, *but*:

- This is solving a problem with how options are stored (i.e. not in a way that properly serializes/deserializes BSON values), not with how we connect
- It makes sense to me to keep the `$compass.xxx` special field handling as localized as possible, not spread across packages

The ideal way to handle this would probably be adjusting the connection storage to handle BSON data well, e.g. by using EJSON instead of JSON, but that might not be trivial (since some of the data might *not* be handled well by EJSON, and we’re aiming to move the storage layer away from Ampersand entirely at some point).